specs/architecture.md

```diff
@@ -0,0 +1,30 @@
+Architecture
+===
+
+- [Architecture](#architecture)
+- [Common Terms and Expressions](#common-terms-and-expressions)
+- [System Architecture](#system-architecture)
+
+# Common Terms and Expressions
+
+| name              | description                                                                                                                                                                      |
+| ----------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| App (application) | Alternate name for "[virtual sidechain](https://arxiv.org/abs/1905.09274)." LazyLedger apps are sidechains that post all their data onto the LazyLedger chain to share security. |
+| Transaction       | Requests that modify the consensus-critical state (validator balances and statuses).                                                                                             |
+| Message           | Requests that are executed by non-consensus-critical apps.                                                                                                                       |
+
+# System Architecture
+
+![fig: Block data structures.](figures/block_data_structures.svg)
+
+LazyLedger has a minimal state: the validator set (account balances, validator status, etc.). Changes to the validator set are done with native _transactions_, distinct from the _messages_ processed by apps. Transactions are signed and must be processed by clients to determine the validator set, while messages are un-signed data blobs that will usually represent an app's block data.
+
+Transactions pay fees similarly to how they would in a normal blockchain (e.g. Bitcoin), and their side effects are restricted to modifying the validator set and their balances. Transactions can additionally pay fees for the inclusion of a message (identified by a hash) in the same block. The validator set is committed to in the block header, and since the entire system state _is_ the validator set, this is the only state commitment needed in the header.
```

Comment:
> So to send a message, I would first need to submit a Tx with a reference to the message (which is then included because a fee for it was payed)?

That's the plan with this.

> Is it necessary that inclusion happens in the same block?

I would say so, yes. 1) You can't have messages in a previous block, obviously, as validators can't see the future. 2) If you allow messages to be included in _future_ blocks, you need to have some sort of guarantee they will be processed, which adds overhead and complexity, opens up the door to DoS vectors, etc. Must simpler to just have whoever is assembling the block be required to make sure that any tx that pays a fee for a message also has that message attached in the same block.

Reply:
> Must simpler to just have whoever is assembling the block be required to make sure that any tx that pays a fee for a message also has that message attached in the same block.

What if whoever is assembling the block didn't receive this message? Does the sender of the Tx receive their paid fee back?

Reply:
There's two alternatives:
1. Such a tx is invalid, so it won't be included in a valid block.
1. The tx sender gets their fee back.

I prefer the former because it involves less work. What do you think?

Reply:
Ok, before finalizing the block going through the Txs and checking if their corresponding message was also received and reject those whose message wasn't, seems much easier. So I would also prefer that (1.). My only concern is that one could effectively DoS the system by sending Tx without a message attached (as the mempool fills up with invalid Tx and there is no or very little cost in doing so).

Reply:
It should be easy to mitigate this DoS vector: just evict transactions that pay for a message unless they're accompanied by a message immediately. Or rather, don't even accept txs that pay for a message into the mempool unless the message is also provided.

Reply:
Oh, it sounded like this were actually two steps. If it is one step and the message is sent along with the Tx (and not only a reference to the message as my original question was implying), we should be fine.

Reply:
And this is good as this exactly how Cosmos SDK transactions work; they are a collection of messages: https://docs.cosmos.network/master/intro/sdk-design.html